src/vm/eventpipe.cpp

```diff
@@ -193,36 +190,49 @@ void EventPipe::Shutdown()
         NOTHROW;
         GC_TRIGGERS;
         MODE_ANY;
+        PRECONDITION(s_pSessions != NULL);
     }
     CONTRACTL_END;
 
+    // TODO: It feels like a Crst is needed here too.
+    CrstHolder _crst(GetLock());
+
     // Mark tracing as no longer initialized.
     s_tracingInitialized = false;
 
+
+    EventPipeConfiguration *pConfig = s_pConfig;
+    EventPipeSessions *pSessions = s_pSessions;
+
+    // Set the static pointers to NULL so that the rest of the EventPipe knows that they are no longer available.
+    // Flush process write buffers to make sure other threads can see the change.
+    s_pConfig = NULL;
+    s_pSessions = NULL;
+
     // We are shutting down, so if disabling EventPipe throws, we need to move along anyway.
     EX_TRY
     {
-        Disable((EventPipeSessionID)s_pSession);
+        if (pSessions != NULL)
+        {
+            // On Ubuntu there is a compilation error because ForEach expects an lvalue.
+            auto functor = [=](EventPipeSession *pSession) {
+                Disable(reinterpret_cast<EventPipeSessionID>(pSession));
+                delete pSession;
+            };
+            pSessions->ForEach(functor);
+        }
     }
     EX_CATCH {}
     EX_END_CATCH(SwallowAllExceptions);
 
-    // Save pointers to the configuration and buffer manager.
-    EventPipeConfiguration *pConfig = s_pConfig;
-    EventPipeBufferManager *pBufferManager = s_pBufferManager;
-
-    // Set the static pointers to NULL so that the rest of the EventPipe knows that they are no longer available.
-    // Flush process write buffers to make sure other threads can see the change.
-    s_pConfig = NULL;
-    s_pBufferManager = NULL;
     FlushProcessWriteBuffers();
```

Comment:
FlushProcessWriteBuffer() should happen right after setting the pointers to NULL?